src/controllers/defiPositions/defiPositions.ts

```diff
@@ -113,120 +115,217 @@ export class DefiPositionsController extends EventEmitter {
+
+          const hasErrors = !!state.providerErrors?.length
+          const filteredPrevious = previousPositions.filter(
+            (prev) =>
+              !customPositions.some((c) => lower(c.providerName) === lower(prev.providerName))
+          )
+
+          this.#state[selectedAccountAddr][chain] = {
+            ...state,
             isLoading: false,
-            positionsByProvider: prevPositionsByProvider || [],
-            error: DeFiPositionsError.CriticalError
+            positionsByProvider: [...filteredPrevious, ...customPositions],
+            updatedAt: hasErrors ? state.updatedAt : Date.now(),
+            error
           }
-          console.error(`updatePositions error on ${n.name}`, e)
-        } finally {
-          this.emitUpdate()
         }
-      })
-    )
+      } catch (e) {
+        console.error(`updatePositions error on ${network.name}`, e)
+        this.#state[selectedAccountAddr][chain] = {
+          providerErrors: this.#state[selectedAccountAddr][chain].providerErrors || [],
+          isLoading: false,
+          positionsByProvider: previousPositions || [],
+          error: DeFiPositionsError.CriticalError
+        }
+      }
+
+      const positionsByProvider = debankPositionsByProvider.filter(
+        (p) => String(p.chainId) === String(network.chainId)
+      )
+
+      const positionMap = new Map(positionsByProvider.map((p) => [lower(p.providerName), p]))
+
+      // eslint-disable-next-line no-restricted-syntax
+      for (const custom of customPositions) {
+        const key = lower(custom.providerName)
+
+        if (custom.providerName === 'Uniswap V3') {
+          const debankUni = positionMap.get(key)
+          if (debankUni) {
+            const merged = {
+              ...debankUni,
+              positions: debankUni.positions.map((pos) => {
+                const match = custom.positions.find(
+                  (p) => p.id === pos.additionalData.positionIndex
+                )
+                return match
+                  ? {
+                      ...pos,
+                      additionalData: {
+                        ...pos.additionalData,
+                        inRange: match.additionalData.inRange
+                      }
+                    }
+                  : pos
+              })
+            } as PositionsByProvider
+
+            positionMap.set(key, merged)
+            // eslint-disable-next-line no-continue
+            continue
+          }
+        }
+
+        positionMap.set(key, custom)
+      }
+
+      this.#state[selectedAccountAddr][chain] = {
+        providerErrors: this.#state[selectedAccountAddr][chain].providerErrors || [],
+        isLoading: false,
+        positionsByProvider: Array.from(positionMap.values()),
+        updatedAt: Date.now()
+      }
+    }
+
+    prepareNetworks()
+
+    try {
+      const resp = await this.#fetch(`https://cena.ambire.com/api/v3/defi/${selectedAccountAddr}`)
```

Comment:
Update: problem confirmed (chatting offline with Simo), if this fails, it should fallback to the custom defi positions, but it doesn't.